src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,847 @@
+                                                                                      s_localizableTitle,
+                                                                                      s_localizableUnsupportedOsVersionMessage,
+                                                                                      DiagnosticCategory.Interoperability,
+                                                                                      RuleLevel.BuildWarning,
+                                                                                      description: s_localizableDescription,
+                                                                                      isPortedFxCopRule: false,
+                                                                                      isDataflowRule: false);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RequiredOsRule, RequiredOsVersionRule, ObsoleteOsRule, UnsupportedOsRule, UnsupportedOsVersionRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var typeName = WellKnownTypeNames.SystemOperatingSystem;
+
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(typeName + "Helper", out var operatingSystemType))
+                {
+                    // TODO: remove 'typeName + "Helper"' after tests able to consume the real new APIs
+                    operatingSystemType = context.Compilation.GetOrCreateTypeByMetadataName(typeName);
+                }
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesOSPlatform, out var osPlatformType) ||
+                    !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesRuntimeInformation, out var runtimeInformationType))
+                {
+                    return;
+                }
+
+                var guardMethods = GetRuntimePlatformGuardMethods(runtimeInformationType, operatingSystemType!);
+
+                context.RegisterOperationBlockStartAction(context => AnalyzeOperationBlock(context, guardMethods, osPlatformType));
+            });
+
+            static ImmutableArray<IMethodSymbol> GetRuntimePlatformGuardMethods(INamedTypeSymbol runtimeInformationType, INamedTypeSymbol operatingSystemType)
+            {
+                return operatingSystemType.GetMembers().OfType<IMethodSymbol>().Where(m => s_platformCheckMethodNames.Contains(m.Name)).ToImmutableArray().
+                    Add(runtimeInformationType.GetMembers().OfType<IMethodSymbol>().Where(m => IsOSPlatform == m.Name).FirstOrDefault());
```

Comment:
I think we should also check for expected return type of the methods (special type boolean) and expected parameter count/type of each method.

Reply:
Agree with the return type, but expected parameter count/type is not fixed though some has no parameter, some has 2, 3 4 parameters some string etc